tortoise/ninja_tortoise.go

```diff
@@ -106,10 +108,45 @@ func NewNinjaTortoise(layerSize int, blocks BlockCache, log log.Log) *ninjaTorto
 	}
 }
 
-func (ni *ninjaTortoise) processBlock(b *types.Block) {
+func (ni *ninjaTortoise) evictOutOfPbase(old types.LayerID) {
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	for i := old; i < ni.pBase.Layer(); i++ {
+		go func() {
+			defer wg.Done()
+			for _, i := range ni.patterns[i] {
+				delete(ni.tSupport, i)
+				delete(ni.tComplete, i)
+				delete(ni.tEffectiveToBlocks, i)
+				delete(ni.tVote, i)
+				delete(ni.tTally, i)
+				delete(ni.tPattern, i)
+				delete(ni.tPatSupport, i)
+				delete(ni.tSupport, i)
+				ni.Debug("evict pattern %v from maps ", i)
+			}
+		}()
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			ids, err := ni.LayerBlockIds(i)
+			if err != nil {
+				ni.Error("could not get layer ids for layer ", i, err)
+			}
+			for _, i := range ids {
+				delete(ni.tEffective, i)
+				delete(ni.tCorrect, i)
+				delete(ni.tExplicit, i)
+				ni.Debug("evict block %v from maps ", i)
+			}
+		}()
+		wg.Wait()
+	}
+}
 
-	ni.Debug("process block: %d layer: %d  ", b.Id, b.Layer())
+func (ni *ninjaTortoise) processBlock(b *types.Block) {
 
+	ni.Debug("process block: %d layer: %d  ", b.ID(), b.Layer())
 	if b.Layer() == Genesis {
 		return
 	}
```

Comment:
Comment regarding old code below - in the `ni.Error` few lines below, you should also print the error